core/cap-0052.md

```diff
@@ -0,0 +1,194 @@
+done in contract code under the limits known today, relatively efficiently.
+
+For example, a single small encode of 32 bytes to 43 bytes is about 200,000
+instructions and about 460 bytes of additional code.
+
+Even though base64 encoding can be done relatively easily for the case of small
+data, in applications that need to encode or decode larger amounts of data it
+quickly become a larger percentage of an applications work.
+
+Additionally, even if efficient, this is a common standard that is simple to
+encode host side and available in many standard libraries today.
+
+An expected use of the base64 encode/decode is in the implementation of a
+webauthn contract, where base64 url encoding is used.
+
+### Goals Alignment
+
+This CAP is aligned with the following Stellar Network Goals:
+
+- The Stellar Network should make it easy for developers of Stellar projects to
+  create highly usable products
+
+## Abstract
+
+This proposal adds two functions to the Soroban environment's exported interface
+that accepts the parameters for configuring a variant of base64, and encode and
+decode bytes to and from base64 as specified in [RFC4648].
+
+## Specification
+
+A new function `base64_encode` with export name `n` in module `b` is added to
+the Soroban environment's exported interface. It accepts an `alphabet`,
+`padding`, and returns the bytes encoded in base64.
+
+Another new function `base64_decode` with export name `o` in module `b` is added
+to the Soroban environment's exported interface. It accepts an `alphabet`,
+`padding`, and returns the bytes decoded. If the decode fails for any reason it
+returns an `ScError`.
+
+For both functions the parameters behave the same.
+
+Parameter `alphabet` specifies the alphabet used by the encoder and decoder. The
+symbols `std` or `url` are accepted, where they map respectively to the standard
+alphabet and the URL and filename friendly alphabet as specified in [RFC4648].
+
+Parameter `padding` specifies the padding character with the same limitations on
+valid values as specified in [RFC4648]. The padding character is 8-bit, and is
+stored in the lower 8-bits of the U32Val type. If the padding value of
+0xffffffff is used it indicates no padding is to be used when encoding, and no
+padding is required when decoding, but padding is permitted. If the padding
+value of 0xfffffffe is used it indicates decoding should fail if the trailing
+padding bits are not zero.
+
+```diff mddiffcheck.ignore=true mddiffcheck.base=v20.1.0
+diff --git a/soroban-env-common/env.json b/soroban-env-common/env.json
+index df7d5c4..ba82167 100644
+--- a/soroban-env-common/env.json
++++ b/soroban-env-common/env.json
+@@ -1890,6 +1890,38 @@
+           ],
+           "return": "U32Val",
+           "docs": "Return the index of a Symbol in an array of linear-memory byte-slices, or trap if not found."
++        },
++        {
++          "export": "n",
++          "name": "base64_encode",
++          "args": [
++            {
```

Comment:
It should include the `input: BytesObject`? Same for the decode function.

Reply:
Ah good catch. I'll add.

Reply:
Fixed thanks.